src/user-data.ts

```diff
@@ -131,3 +130,196 @@ export type AppSemester = {
   readonly type: FirestoreSemesterType;
   readonly year: number;
 };
+
+/**
+ * Creates credit range based on course
+ * Example: [1, 4] is the credit range for the given course
+ */
+const createCourseCreditRange = (course: FirestoreSemesterCourse): readonly [number, number] => {
+  const courseCreditRange: number[] = [];
+  if (typeof course.creditRange !== 'undefined') {
+    return course.creditRange;
+  }
+  if (typeof course.enrollGroups !== 'undefined') {
+    course.enrollGroups.forEach(enrollGroup => {
+      courseCreditRange.push(enrollGroup.unitsMinimum);
+      courseCreditRange.push(enrollGroup.unitsMaximum);
+    });
+    return [Math.min(...courseCreditRange), Math.max(...courseCreditRange)];
+  }
+  // @ts-ignore
+  return [course.credits, course.credits];
+};
+
+export const firestoreCourseToAppCourse = (
+  course: FirestoreSemesterCourse,
+  incrementID: () => number,
+  addColor: (subject: string) => string
```

Comment:
This is very sad because the old conversion code is not pure functions, but with some tricky side effects tightly coupled with the component, so I think the only way is to make the caller explicitly pass in those side effect functions.